### ISSUE TITLE
Proposal: Enforce consistent linebreak style for semicolons

### ISSUE THREAD

**Issue by paulmelnikow** [eslint/eslint#8169]
**Please describe what the rule should do:**

Enforce consistent linebreak style for semicolons. Require that they be placed either at the end of a line or beginning of a line.

While in a project that use semicolons after every statement, there is little question that the semicolon belongs at the end of the line. However, in a project that uses semicolons only where needed, it's helpful to enforce the placement of the occasional semicolon.

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

With `"after"`, this would be an error:

```js
foo()
bar(); // A semicolon is required here for the next statement to be interpreted as intended.
[1, 2, 3].forEach(a => { console.log(a) })
baz()
```

This would be correct:

```js
foo()
bar() // A semicolon is required here for the next statement to be interpreted as intended.
;[1, 2, 3].forEach(a => { console.log(a) })
baz()
```

With `"before"`, this would be an error:

```js
foo();
bar()
;[1, 2, 3].forEach(a => { console.log(a) });
baz();
```

This would be correct:

```js
foo();
bar();
[1, 2, 3].forEach(a => { console.log(a) });
baz();
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Semicolon style is applicable to most projects that define a style guide. It varies significantly from project to project. Some using semis everywhere; others only where they're needed. Developers new to no-semicolon projects may misplace the required semicolon. Developers on semicolon projects may prefer to enforce consistency as well.

This rule relates to `semi`, but does not depend on, conflict with, or overlap with it, or any other rule.

The rule is straightforward to understand. It's just like `operator-linebreak`.

**Comment by mysticatea** [comment]
Thank you for this issue.

Sounds reasonable to me, since we have similar rules; [comma-style] and [operator-linebreak].
It would be useful when somebody want to switch semicolon-style by autofix.

(The options would close to [comma-style]'s rather than `before`/`after`.)

[comma-style]: http://eslint.org/docs/rules/comma-style
[operator-linebreak]: http://eslint.org/docs/rules/operator-linebreak

**Comment by not-an-aardvark** [comment]
Would this conflict with [`semi-spacing`](http://eslint.org/docs/rules/semi-spacing)?

**Comment by mysticatea** [comment]
I think it's the same relation as between [comma-style] and [comma-spacing]. We are doing multipass autofix, so those would apply correctly eventually.

[comma-style]: http://eslint.org/docs/rules/comma-style
[comma-spacing]: http://eslint.org/docs/rules/comma-spacing

**Comment by paulmelnikow** [comment]
@mysticatea Interesting, yes, [comma-style](http://eslint.org/docs/rules/comma-style) seems like a good example to follow. I see it generates a "Bad line breaking before and after ','" message which would be useful for semis. Good point about autofix too.

**Comment by ilyavolodin** [comment]
This rule would only work (and make sense) with `semi: ["error", "never"]`. We usually don't do rules that directly conflict with other rules... Not sure if this is a good idea.

**Comment by not-an-aardvark** [comment]
Isn't the following also valid with `semi: ["error", "always"]`?

```js
foo()
; bar()
; baz()
;
```

**Comment by ilyavolodin** [comment]
@not-an-aardvark Well, yes, it is. But I've never seen anyone use that style.

**Comment by paulmelnikow** [comment]
`semi: ["error", "always"]`, `semi-spacing`, `indent` – none of these will reject @not-an-aardvark's snippet, but this rule would. (`semi-style: ["error", "last"]`?)

The same is true of the misplaced semicolon in this one:
```js
const a = 1

;
```

**Comment by paulmelnikow** [comment]
@ilyavolodin While one option makes more sense with `semi: ["error", "always"]` and another with `semi: ["error", "never"]`, the rule catches errors in both cases. Does that make sense?

**Comment by ilyavolodin** [comment]
@paulmelnikow It does as a statement, but as a rule, we have no precedence for such tightly coupled rules. I'm still not convinced that it would be a good idea to go down this route.

**Comment by platinumazure** [comment]
I don't agree that the rule is coupled to `"semi": ["error", "never"]`. It might be more sensible with that configuration, but it also could work with semicolon-required style (even if that's not a common style). But the rules don't actually know about each other or depend on each other's configuration.

I'm not 100% convinced this rule is important enough to be included in core, though. We have a lot of core rules so new rules need to meet a very high bar for inclusion, and I'm not convinced this does.

**Comment by mysticatea** [comment]
I'd like to champion this. I think this fits to core.

- There were enhancement requests for linebreaks around semicolons before. For example, `semi-style` rule probably covers #7954.
- We have similar rule `comma-style`.

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team We need one more :+1: from a team member to accept this proposal. Is anyone willing to support it?

**Comment by kaicataldo** [comment]
This has now been accepted!

**Comment by paulmelnikow** [comment]
@mysticatea I didn't realize you were working on this already! Awesome! 😁
